node/pkg/accountant/submit_obs.go

```diff
@@ -106,60 +169,113 @@ func (sb SignatureBytes) MarshalJSON() ([]byte, error) {
-		pe, exists := acct.pendingTransfers[msgId]
-		if exists {
-			acct.logger.Info("acct: transfer has already been committed, publishing it", zap.String("msgId", msgId))
-			acct.publishTransfer(pe)
-			transfersApproved.Inc()
-		} else {
-			acct.logger.Debug("acct: transfer has already been committed, and it is no longer in our map", zap.String("msgId", msgId))
+	if len(responses) != len(msgs) {
+		// This means the whole batch failed. They will all get retried the next audit cycle.
+		acct.logger.Error("acct: number of responses does not match number of messages", zap.Int("numMsgs", len(msgs)), zap.Int("numResp", len(responses)), zap.Error(err))
+		for idx, msg := range msgs {
+			acct.logger.Error("acct: need to retry observation", zap.Int("idx", idx), zap.String("msgId", msg.MessageIDString()))
 		}
+
+		submitFailures.Add(float64(len(msgs)))
+		return
+	}
+
+	for _, msg := range msgs {
+		msgId := msg.MessageIDString()
+
+		status, exists := responses[msgId]
+		if !exists {
+			// This will get retried next audit interval.
+			acct.logger.Error("acct: did not receive an observation response for message", zap.String("msgId", msgId))
+			submitFailures.Inc()
+			continue
+		}
+
+		switch status.Type {
+		case "pending":
+			acct.logger.Info("acct: transfer is pending", zap.String("msgId", msgId))
+		case "committed":
+			acct.handleCommittedTransfer(msgId)
+		case "error":
+			submitFailures.Inc()
+			acct.handleTransferError(msgId, status.Data, "acct: transfer failed")
+		default:
+			// This will get retried next audit interval.
+			acct.logger.Error("acct: unexpected status response on observation", zap.String("msgId", msgId), zap.String("status", status.Type), zap.String("text", status.Data))
+			submitFailures.Inc()
+		}
+	}
+}
+
+// handleCommittedTransfer updates the pending map and publishes a committed transfer. It grabs the lock.
+func (acct *Accountant) handleCommittedTransfer(msgId string) {
+	acct.pendingTransfersLock.Lock()
+	defer acct.pendingTransfersLock.Unlock()
+	pe, exists := acct.pendingTransfers[msgId]
+	if exists {
+		acct.logger.Info("acct: transfer has already been committed, publishing it", zap.String("msgId", msgId))
+		acct.publishTransferAlreadyLocked(pe)
+		transfersApproved.Inc()
+	} else {
+		acct.logger.Debug("acct: transfer has already been committed but it is no longer in our map", zap.String("msgId", msgId))
+	}
+}
+
+// handleTransferError is called when a transfer fails, either from a submit or an event notification. It handles insufficient balance error. It grabs the lock.
+func (acct *Accountant) handleTransferError(msgId string, errText string, logText string) {
+	if strings.Contains(errText, "insufficient balance") {
+		balanceErrors.Inc()
+		acct.logger.Error("acct: insufficient balance error detected, dropping transfer", zap.String("msgId", msgId), zap.String("text", errText))
+		acct.deletePendingTransfer(msgId)
+	} else {
+		// This will get retried next audit interval.
```

Comment:
I think we should also have a metric for non-"insufficient balance" errors as well.  Now that a failed observation doesn't fail the whole batch, errors should not be expected during normal operation.

I'm fine with adding this in a follow-up if you want to merge this PR first.

Reply:
I think I handle this already by either pegging "submit failures" or "error events", depending on where handleTransferError() is called from.